File: tests/gpu/uniques_test.py
"""THE UNIQUE UNITS on the GPU engine — the twin of tests/cpu/units/uniques.test.ts.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/uniques_test.py

Checks:
  A. `row_civ` reads the fixture's `leader`; `_civ_unit_ok` hands each unique
     to its civilization and removes the chassis it replaces there; a
     city-state row trains no unique.
  B. `_up_to_row` lands the upgrade on the civilization's replacement and
     `_seat_trainable_units` / the production mask follow `_civ_unit_ok`.
  C. the chariot classes are cavalry but not an anti-cavalry target; the
     Heavy Chariot halts in enemy ZOC where the War-Cart, the Maryannu and the
     Longship walk through.
  D. `_start_tile_mp`: open flat ground pays the chariots, hills and snow do
     not; the Berserker draws +2 in enemy territory only at war; the Longship
     +1 on coast only.
  E. Berserker Rage: `_chassis_atk_cs_pool` reads +10, `_type_def_melee_cs` -5.
  F. the Legion is military with one charge, lays the FORT column on the
     engineer's ground without the Fort's tech, and outlives the charge.
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
from core import load_rules, fixture_paths  # noqa: E402
from engineer_test import fresh, retype, mask_of, order, clear_tile, own_flat, neutral_flat  # noqa: E402


def stand(sim, row, ty, tile):
    """A live row-`row` unit retyped to `ty` on `tile`, registered on the
    occupancy plane its domain owns."""
    v = retype(sim, row, ty, tile)
    lo = sim.POOL_LO["major"]
    sim.civilian_at[0, tile] = -1
    sim.military_at[0, tile] = -1
    if bool(sim._type_civilian[ty]):
        sim.civilian_at[0, tile] = v + lo
    else:
        sim.military_at[0, tile] = v + lo
    sim._gen_ver += 1
    return v


def main() -> None:
    rules = load_rules()
    path = fixture_paths()[0]
    U = {u["id"]: i for i, u in enumerate(rules.units)}
    C = {c: i for i, c in enumerate(rules.uniques["civs"])}
    sim = fresh(rules, path)
    R = sim.n_majors
    rome = next(r for r in range(R) if int(sim.row_civ[r]) == C["ROME"])
    egypt = next(r for r in range(R) if int(sim.row_civ[r]) == C["EGYPT"])
    norway = next(r for r in range(R) if int(sim.row_civ[r]) == C["NORWAY"])
    assert all(int(sim.row_civ[r]) >= 0 for r in range(R)), "every major plays a civilization"
    cs_row = R
    assert int(sim.row_civ[cs_row]) < 0, "a city-state row plays none"

    # -- A: who may train what ---------------------------------------------
    def ok(row, uid):
        return bool(sim._civ_unit_ok(row)[U[uid]])

    assert ok(rome, "LEGION") and not ok(rome, "SWORDSMAN")
    assert ok(egypt, "SWORDSMAN") and not ok(egypt, "LEGION")
    assert ok(egypt, "MARYANNU_CHARIOT_ARCHER") and not ok(rome, "MARYANNU_CHARIOT_ARCHER")
    assert ok(norway, "LONGSHIP") and not ok(norway, "GALLEY")
    assert ok(norway, "BERSERKER") and not ok(norway, "MAN_AT_ARMS")
    assert ok(rome, "GALLEY") and ok(rome, "MAN_AT_ARMS")
    assert not ok(rome, "WAR_CART") and not ok(norway, "WAR_CART")
    assert ok(cs_row, "SWORDSMAN") and ok(cs_row, "GALLEY")
    for uid in ("LEGION", "MARYANNU_CHARIOT_ARCHER", "BERSERKER", "LONGSHIP", "WAR_CART"):
        assert not ok(cs_row, uid), f"a city-state row trains no unique ({uid})"
    print("  A civ_unit_ok OK")

    # -- B: the upgrade target and the two trainable masks -----------------
    assert int(sim._up_to_row(norway)[U["SWORDSMAN"]]) == U["BERSERKER"]
    assert int(sim._up_to_row(rome)[U["SWORDSMAN"]]) == U["MAN_AT_ARMS"]
    assert int(sim._up_to_row(rome)[U["WARRIOR"]]) == U["LEGION"]
    assert int(sim._up_to_row(norway)[U["LEGION"]]) == U["BERSERKER"]
    assert int(sim._up_to_row(cs_row)[U["SWORDSMAN"]]) == U["MAN_AT_ARMS"]
    assert int(sim._up_to_row(norway)[U["QUADRIREME"]]) == int(sim._type_up_to[U["QUADRIREME"]])
    wheel = int(sim._type_tech[U["MARYANNU_CHARIOT_ARCHER"]])
    assert wheel >= 0
    for r in (egypt, rome):
        sim.civ_techs[0, r, wheel] = True
    sim._gen_ver += 1
    assert bool(sim._seat_trainable_units(egypt)[0, U["MARYANNU_CHARIOT_ARCHER"]])
    assert not bool(sim._seat_trainable_units(rome)[0, U["MARYANNU_CHARIOT_ARCHER"]])
    assert bool(sim._seat_trainable_units(rome)[0, U["HEAVY_CHARIOT"]])
    pm_e = sim._trainable_units(egypt)
    pm_r = sim._trainable_units(rome)
    assert bool(pm_e[..., U["MARYANNU_CHARIOT_ARCHER"]].any()), "Egypt's production mask offers the Maryannu"
    assert not bool(pm_r[..., U["MARYANNU_CHARIOT_ARCHER"]].any()), "Rome's production mask does not"
    assert not bool(pm_r[..., U["SWORDSMAN"]].any()), "Rome never sees the Swordsman"
    print("  B upgrade target + trainable masks OK")

    # -- C: chariots vs the anti-cavalry bonus, and ZOC --------------------
    def cs(a, d):
        return int(sim._class_matchup_cs(torch.tensor([U[a]]), torch.tensor([U[d]]))[0])

    assert cs("SPEARMAN", "HORSEMAN") == 10
    assert cs("SPEARMAN", "HEAVY_CHARIOT") == 0
    assert cs("SPEARMAN", "WAR_CART") == 0
    assert cs("SPEARMAN", "MARYANNU_CHARIOT_ARCHER") == 10
    assert bool(sim._type_cavalry[U["HEAVY_CHARIOT"]]) and bool(sim._type_cavalry[U["WAR_CART"]])
    t_foe = own_flat(sim, egypt)
    clear_tile(sim, t_foe)
    stand(sim, egypt, U["WARRIOR"], t_foe)
    dest = next(int(x) for x in sim.neigh[t_foe].tolist()
                if x >= 0 and not bool(sim.water[0, x]) and bool(sim.passable[0, x]))
    ra, rb = sim._seat_row[rome], sim._seat_row[egypt]
    sim.war[0, ra, rb] = sim.war[0, rb, ra] = True
    sim._gen_ver += 1

    def zoc(uid):
        return bool(sim._in_enemy_zoc(torch.tensor([dest]), rome, torch.tensor([U[uid]]))[0])

    assert zoc("HEAVY_CHARIOT"), "a Heavy Chariot halts in enemy ZOC"
    assert not zoc("HORSEMAN")
    assert not zoc("WAR_CART"), "CLASS_WAR_CART ignores ZOC"
    assert not zoc("MARYANNU_CHARIOT_ARCHER"), "CLASS_RANGED_CAVALRY ignores ZOC"
    assert not zoc("LONGSHIP"), "CLASS_LONGSHIP ignores ZOC"
    print("  C chariot matchup + ZOC OK")

    # -- D: the start-tile Movement ---------------------------------------
    mp = sim._mp_scale
    open_t = [int(t) for t in sim._open_terr.tolist()]
    flat = next(t for t in range(sim.T)
                if int(sim.terrain[0, t]) in open_t and not bool(sim.hills[0, t])
                and not bool(sim.water[0, t]) and bool(sim.passable[0, t]))
    hill = next(t for t in range(sim.T)
                if int(sim.terrain[0, t]) in open_t and bool(sim.hills[0, t]) and bool(sim.passable[0, t]))
    snow = next((t for t in range(sim.T)
                 if int(sim.terrain[0, t]) not in open_t and not bool(sim.hills[0, t])
                 and not bool(sim.water[0, t]) and bool(sim.passable[0, t])), -1)

    def full(row, uid, tile):
        v = stand(sim, row, U[uid], tile)
        return float(sim._full_mp("major")[0, v]) / mp

    assert full(rome, "HEAVY_CHARIOT", flat) == 3.0
    assert full(rome, "HEAVY_CHARIOT", hill) == 2.0
    if snow >= 0:
        assert full(rome, "HEAVY_CHARIOT", snow) == 2.0
    assert full(egypt, "MARYANNU_CHARIOT_ARCHER", flat) == 4.0
    assert full(rome, "WAR_CART", flat) == 4.0
    assert full(rome, "WARRIOR", flat) == 2.0
    theirs = own_flat(sim, egypt, avoid=(t_foe,))
    assert full(rome, "BERSERKER", theirs) == 4.0, "at war: +2 in enemy territory"
    sim.war[0, ra, rb] = sim.war[0, rb, ra] = False
    sim._gen_ver += 1
    assert full(rome, "BERSERKER", theirs) == 2.0, "at peace: their land is not 'enemy territory'"
    assert full(rome, "BERSERKER", own_flat(sim, rome)) == 2.0
    coast = next(t for t in range(sim.T) if int(sim.terrain[0, t]) == sim._coast_terr)
    ocean = next((t for t in range(sim.T)
                  if bool(sim.water[0, t]) and int(sim.terrain[0, t]) != sim._coast_terr), -1)
    assert full(norway, "LONGSHIP", coast) == 4.0, "+1 Movement while in coastal waters"
    assert full(norway, "GALLEY", coast) == 3.0
    if ocean >= 0:
        assert full(norway, "LONGSHIP", ocean) == 3.0
    print("  D start-tile Movement OK")

    # -- E: Berserker Rage ------------------------------------------------
    v = stand(sim, norway, U["BERSERKER"], own_flat(sim, norway))
    assert int(sim._chassis_atk_cs_pool("major", torch.tensor([v]))[0]) == 10
    assert int(sim._type_def_melee_cs[U["BERSERKER"]]) == -5
    v2 = stand(sim, norway, U["MAN_AT_ARMS"], own_flat(sim, norway))
    assert int(sim._chassis_atk_cs_pool("major", torch.tensor([v2]))[0]) == 0
    assert int(sim._type_def_melee_cs[U["MAN_AT_ARMS"]]) == 0
    print("  E Berserker Rage OK")

    # -- F: the Legion's Fort ---------------------------------------------
    sim = fresh(rules, path)
    FORT = sim.FORT
    assert FORT >= 0 and sim._imp_eng[FORT]
    assert not bool(sim._type_civilian[U["LEGION"]]) and int(sim._type_charges[U["LEGION"]]) == 1
    assert bool(sim._type_fort_builder[U["LEGION"]])
    t = own_flat(sim, rome)
    clear_tile(sim, t)
    v = stand(sim, rome, U["LEGION"], t)
    ut = int(sim._imp_unlock[FORT])
    if ut >= 0:
        sim.civ_techs[0, rome, ut] = False
    sim._gen_ver += 1
    um = mask_of(sim, rome, v)
    c_fort = sim._A_IMP[FORT]
    assert bool(um[c_fort]), "a Legion on bare own flat land is offered the Fort, no tech needed"
    others = [k for k, e in enumerate(sim._imp_eng) if e and k != FORT and sim._A_IMP[k] >= 0]
    for k in others:
        assert not bool(um[sim._A_IMP[k]]), "the Legion lays the Fort alone, not the engineer's other rows"
    v_sw = stand(sim, rome, U["SWORDSMAN"], t)
    assert v_sw == v
    assert not bool(mask_of(sim, rome, v_sw)[c_fort]), "a Swordsman lays nothing"
    stand(sim, rome, U["LEGION"], t)
    t_neu = neutral_flat(sim, avoid=(t,))
    clear_tile(sim, t_neu)
    stand(sim, rome, U["LEGION"], t_neu)
    assert bool(mask_of(sim, rome, v)[c_fort]), "the engineer's ground: neutral tiles too"
    order(sim, rome, v, c_fort)
    assert int(sim.improvement[0, t_neu]) == FORT, "the Fort was not placed"
    assert int(sim.major_unit_charges[0, v]) == 0, "the charge is spent"
    assert float(sim.major_unit_mp[0, v]) == 0.0, "the turn is spent"
    assert bool(sim.major_unit_alive[0, v]), "a military chassis outlives its last charge"
    assert not bool(mask_of(sim, rome, v)[c_fort]), "no charge, no second Fort"
    print("  F Legion Fort OK")
    print("UNIQUES OK")


if __name__ == "__main__":
    main()

File: tests/cpu/units/uniques.test.ts
import { describe, it, expect } from 'vitest';
import { MP_SCALE } from '../../../cpu/data/constants';
import { makeMap, makeState, tileAtCoords } from '../helpers';
import { CIV_IDS, CIV_LEADERS } from '../../../cpu/data/seats';
import { UNITS, civUnitAllowed, civReplacement, civUpgradeTarget } from '../../../cpu/data/units';
import { civOf, emptySeat, setTileOwner, setWar } from '../../../cpu/core/seats';
import { inEnemyZoc, spawnUnit, startTileMoves, trainableUnits, unitDomain, unitFullMoves } from '../../../cpu/core/units';
import { chassisAttackCS, chassisDefendCS, classMatchupCS, defenderCS } from '../../../cpu/core/combat';
import { upgradeResourceCost } from '../../../cpu/core/stockpile';
import { validImprovementsIn } from '../../../cpu/core/rules';

/**
 * THE UNIQUE UNITS (CIV6, Civilizations.xml / Units.xml of the owner's
 * install): each civilization's own chassis, what it stands in for, and the
 * terms only a unique carries. Gate reachability is thin — the fixtures seat
 * Rome, Egypt and Norway and never Sumeria — so the rules are pinned here.
 */
describe('the seat plays a civilization', () => {
  it('CIV_LEADERS and CIV_IDS agree, and civOf reads the seat', () => {
    // one row per civilization-LEADER pair: every row names a civilization
    // the list holds, every civilization has a row, and the first four rows
    // are the developed four in `CIV_IDS` order
    for (const l of CIV_LEADERS) expect(CIV_IDS).toContain(l.civ);
    for (const c of CIV_IDS) expect(CIV_LEADERS.some((l) => l.civ === c)).toBe(true);
    CIV_LEADERS.slice(0, 4).forEach((l, i) => expect(l.civ).toBe(CIV_IDS[i]));
    expect(new Set(CIV_LEADERS.map((l) => l.leader)).size).toBe(CIV_LEADERS.length);
    const state = makeState(makeMap(8, 8));
    expect(civOf(state, 0)).toBeNull(); // a bare seat plays none
    state.seats[0].civ = 0;
    expect(civOf(state, 0)).toBe('ROME');
    state.seats[0].civ = 2;
    expect(civOf(state, 0)).toBe('NORWAY');
  });

  it('a unique trains for its civilization alone and replaces its base chassis there', () => {
    expect(civUnitAllowed('ROME', 'LEGION')).toBe(true);
    expect(civUnitAllowed('ROME', 'SWORDSMAN')).toBe(false);
    expect(civUnitAllowed('EGYPT', 'SWORDSMAN')).toBe(true);
    expect(civUnitAllowed('EGYPT', 'LEGION')).toBe(false);
    expect(civUnitAllowed('EGYPT', 'MARYANNU_CHARIOT_ARCHER')).toBe(true);
    expect(civUnitAllowed('NORWAY', 'LONGSHIP')).toBe(true);
    expect(civUnitAllowed('NORWAY', 'GALLEY')).toBe(false);
    expect(civUnitAllowed('NORWAY', 'BERSERKER')).toBe(true);
    expect(civUnitAllowed('NORWAY', 'MAN_AT_ARMS')).toBe(false);
    expect(civUnitAllowed('SUMERIA', 'WAR_CART')).toBe(true);
    expect(civUnitAllowed('ROME', 'WAR_CART')).toBe(false);
    // a seat playing no civilization trains every base chassis and no unique
    expect(civUnitAllowed(null, 'SWORDSMAN')).toBe(true);
    expect(civUnitAllowed(null, 'LEGION')).toBe(false);
    expect(civReplacement('ROME', 'SWORDSMAN')).toBe('LEGION');
    expect(civReplacement('ROME', 'GALLEY')).toBeUndefined();
  });

  it('trainableUnits follows the civilization', () => {
    const state = makeState(makeMap(8, 8));
    state.unitsMode = true;
    const ids = () => trainableUnits(state, 0).map((d) => d.id);
    state.seats[0].civ = 3; // Sumeria: the War-Cart needs no tech
    expect(ids()).toContain('WAR_CART');
    state.seats[0].civ = 0;
    expect(ids()).not.toContain('WAR_CART');
    expect(ids()).not.toContain('SWORDSMAN');
    state.seats[0].civ = -1;
    expect(ids()).not.toContain('WAR_CART');
  });

  it('the upgrade lands on the civilization\'s replacement', () => {
    expect(civUpgradeTarget('NORWAY', 'SWORDSMAN')).toBe('BERSERKER');
    expect(civUpgradeTarget('ROME', 'SWORDSMAN')).toBe('MAN_AT_ARMS');
    expect(civUpgradeTarget(null, 'SWORDSMAN')).toBe('MAN_AT_ARMS');
    expect(civUpgradeTarget('ROME', 'WARRIOR')).toBe('LEGION');
    expect(civUpgradeTarget('NORWAY', 'LEGION')).toBe('BERSERKER');
    expect(civUpgradeTarget('NORWAY', 'QUADRIREME')).toBe(UNITS.QUADRIREME.upgradesTo);
    const state = makeState(makeMap(8, 8));
    state.seats[0].civ = 0; // Rome: Warrior -> Legion charges the Legion's Iron
    expect(upgradeResourceCost(state, 0, 'WARRIOR')).toEqual({ id: 'IRON', n: 20 });
    state.seats[0].civ = 2; // Norway: Swordsman (Iron) -> Berserker (Iron) is free
    expect(upgradeResourceCost(state, 0, 'SWORDSMAN')).toBeUndefined();
  });
});

describe('the chariot classes', () => {
  it('are cavalry, but not a target of the anti-cavalry bonus', () => {
    // CIV6 (ANTI_CAVALRY_OPPONENT_REQUIREMENTS): light, heavy and RANGED
    // cavalry — a chariot class is not on the list.
    expect(classMatchupCS('SPEARMAN', 'HORSEMAN')).toBe(10);
    expect(classMatchupCS('SPEARMAN', 'HEAVY_CHARIOT')).toBe(0);
    expect(classMatchupCS('SPEARMAN', 'WAR_CART')).toBe(0);
    expect(classMatchupCS('SPEARMAN', 'MARYANNU_CHARIOT_ARCHER')).toBe(10);
    expect(UNITS.HEAVY_CHARIOT.cavalry).toBe(true);
    expect(UNITS.WAR_CART.cavalry).toBe(true);
  });

  it('the Heavy Chariot halts in enemy ZOC; the War-Cart and the ranged chariot ignore it', () => {
    const state = makeState(makeMap(8, 8));
    state.unitsMode = true;
    state.seats.push({ ...emptySeat(1), name: 'foe' });
    setWar(state, 0, 1, true);
    const foe = tileAtCoords(state.map, 3, 3);
    spawnUnit(state, 'WARRIOR', foe.index, 1);
    const dest = tileAtCoords(state.map, 4, 3);
    expect(inEnemyZoc(state, dest.index, { seat: 0, type: 'HEAVY_CHARIOT' })).toBe(true);
    expect(inEnemyZoc(state, dest.index, { seat: 0, type: 'HORSEMAN' })).toBe(false);
    expect(inEnemyZoc(state, dest.index, { seat: 0, type: 'WAR_CART' })).toBe(false); // CLASS_WAR_CART
    expect(inEnemyZoc(state, dest.index, { seat: 0, type: 'MARYANNU_CHARIOT_ARCHER' })).toBe(false); // CLASS_RANGED_CAVALRY
    expect(inEnemyZoc(state, dest.index, { seat: 0, type: 'LONGSHIP' })).toBe(false); // CLASS_LONGSHIP
  });
});

describe('start-of-turn Movement from the tile', () => {
  it('open flat terrain pays the chariots, hills and forest ground do not', () => {
    const state = makeState(makeMap(8, 8, 'GRASSLAND'));
    const flat = tileAtCoords(state.map, 2, 2);
    const hills = tileAtCoords(state.map, 3, 2);
    hills.elevation = 'HILLS';
    const snow = tileAtCoords(state.map, 4, 2);
    snow.terrain = 'SNOW';
    const u = (type: string, t: { index: number }) => ({ type, seat: 0, tileIndex: t.index });
    expect(startTileMoves(state, u('HEAVY_CHARIOT', flat))).toBe(1);
    expect(startTileMoves(state, u('HEAVY_CHARIOT', hills))).toBe(0);
    expect(startTileMoves(state, u('HEAVY_CHARIOT', snow))).toBe(0);
    expect(startTileMoves(state, u('MARYANNU_CHARIOT_ARCHER', flat))).toBe(2);
    expect(startTileMoves(state, u('WAR_CART', flat))).toBe(1);
    expect(startTileMoves(state, u('WARRIOR', flat))).toBe(0);
    // it rides unitFullMoves and the spawn's first turn
    expect(unitFullMoves(state, u('HEAVY_CHARIOT', flat))).toBe(MP_SCALE * 3);
    expect(unitFullMoves(state, u('HEAVY_CHARIOT', hills))).toBe(MP_SCALE * 2);
    state.unitsMode = true;
    const w = spawnUnit(state, 'WAR_CART', flat.index, 0)!;
    expect(w.movesLeft).toBe(MP_SCALE * 4);
  });

  it('the Berserker draws +2 in enemy territory, the Longship +1 on coast', () => {
    const state = makeState(makeMap(8, 8, 'PLAINS'));
    state.seats.push({ ...emptySeat(1), name: 'foe' });
    const theirs = tileAtCoords(state.map, 2, 2);
    setTileOwner(theirs, 1, 1);
    const b = { type: 'BERSERKER', seat: 0, tileIndex: theirs.index };
    expect(startTileMoves(state, b)).toBe(0); // at peace: not "enemy" territory
    setWar(state, 0, 1, true);
    expect(startTileMoves(state, b)).toBe(2);
    const mine = tileAtCoords(state.map, 3, 2);
    setTileOwner(mine, 0, 1);
    expect(startTileMoves(state, { ...b, tileIndex: mine.index })).toBe(0);
    const coast = tileAtCoords(state.map, 5, 5);
    coast.terrain = 'COAST';
    const ocean = tileAtCoords(state.map, 6, 5);
    ocean.terrain = 'OCEAN';
    expect(startTileMoves(state, { type: 'LONGSHIP', seat: 0, tileIndex: coast.index })).toBe(1);
    expect(startTileMoves(state, { type: 'LONGSHIP', seat: 0, tileIndex: ocean.index })).toBe(0);
    expect(startTileMoves(state, { type: 'GALLEY', seat: 0, tileIndex: coast.index })).toBe(0);
  });
});

describe('Berserker Rage', () => {
  it('+10 attacking, -5 defending against melee only', () => {
    expect(chassisAttackCS({ type: 'BERSERKER' })).toBe(10);
    expect(chassisAttackCS({ type: 'MAN_AT_ARMS' })).toBe(0);
    expect(chassisDefendCS({ type: 'BERSERKER' }, { melee: true })).toBe(-5);
    expect(chassisDefendCS({ type: 'BERSERKER' }, { melee: false })).toBe(0);
    expect(chassisDefendCS({ type: 'BERSERKER' })).toBe(0);
    const state = makeState(makeMap(8, 8, 'PLAINS'));
    state.unitsMode = true;
    state.seats.push({ ...emptySeat(1), name: 'foe' });
    const here = tileAtCoords(state.map, 3, 3);
    const d = spawnUnit(state, 'BERSERKER', here.index, 0)!;
    const a = spawnUnit(state, 'WARRIOR', tileAtCoords(state.map, 4, 3).index, 1)!;
    const melee = defenderCS(state, d, here.index, { attacker: a, melee: true });
    const ranged = defenderCS(state, d, here.index, { attacker: a, melee: false });
    expect(melee - ranged).toBe(-5);
    d.type = 'MAN_AT_ARMS';
    expect(defenderCS(state, d, here.index, { attacker: a, melee: true })
      - defenderCS(state, d, here.index, { attacker: a, melee: false })).toBe(0);
  });
});

describe('the Legion', () => {
  it('is a military unit with a charge, and lays the Fort alone', () => {
    expect(unitDomain('LEGION')).toBe('military');
    expect(unitDomain('BUILDER')).toBe('civilian');
    expect(UNITS.LEGION.charges).toBe(1);
    const t = tileAtCoords(makeMap(8, 8, 'PLAINS'), 2, 2);
    const opts = { unlocks: null, ownsTile: () => true };
    expect(validImprovementsIn(t, { ...opts, builder: 'LEGION' })).toEqual(['FORT']);
    expect(validImprovementsIn(t, { ...opts, builder: 'SWORDSMAN' })).not.toContain('FORT');
    expect(validImprovementsIn(t, { ...opts, builder: 'BUILDER' })).not.toContain('FORT');
    // the engineer's ground, not the builder's: neutral tiles are fine
    expect(validImprovementsIn(t, { unlocks: null, ownsTile: () => false, builder: 'LEGION' })).toEqual(['FORT']);
  });
});
